### DISCUSSION TITLE
If trace sampling, metric can reflect the real situation of the system?

### DISCUSSION THREAD

**Discussion by chenhaipeng** [apache/skywalking#12222] [Q&A] [answered]
I've noticed that Skywalking metrics are generated based on traces. However, since traces are sampled, the metrics are not entirely accurate.

**Comment by wu-sheng** [comment]
Of course. What are you asking about?

**Reply by chenhaipeng** [reply]
If trace sampling leads to inaccurate metrics, then the reference value of the metrics becomes questionable. For instance, if I'm calculating the QPS  of an interface, and trace sampling only records a portion of the requests, then the QPS metrics generated from the traces no longer accurately reflect the actual traffic of the interface.

**Reply by chenhaipeng** [reply]
This will lead to the inability to correctly reflect the real situation of the system.

**Reply by wu-sheng** [reply]
As same. I mean what do you ask for?

Of course, nothing to report, nothing to calculate. 

There is no magic.

**Reply by wu-sheng** [reply]
Server side sampling makes sure the data accurate. 
If you want to use agent sampling, of course, don't expect magic happens. There's not in the real world.

**Reply by chenhaipeng** [reply]
No reporting, no calculation. That's true. If metrics are distorted under trace sampling, and metrics are important indicators for system measurement, especially in transaction scenarios where sampling should not be used, is it a reasonable suggestion for Skywalking to have a custom self-reporting interface for metrics?

**Reply by wu-sheng** [reply, marked as answer]
There are metric plugin(for connection pool, etc.) and toolkit of meter, or you could use Spring lib

- https://skywalking.apache.org/docs/skywalking-java/latest/en/setup/service-agent/java-agent/application-toolkit-meter/
- https://skywalking.apache.org/docs/skywalking-java/latest/en/setup/service-agent/java-agent/application-toolkit-micrometer-1.10/
